### PR TITLE
Maintenance/great 1103/great 1318/alexsherstinsky/validation graph/refactor validation graph usage 2022 10 20 248

### PR DESCRIPTION
**Pull request by alexsherstinsky**
This Pull Request has been created because #6184 was accidentally closed.  Please refer to the discussion therein.  Thanks.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/validator/validation_graph.py [hunk 1]
```diff
@@ -28,20 +40,190 @@ def id(self):
...
+                    for failed_metric in err.failed_metrics:
+                        if failed_metric.id in failed_metric_info:
+                            failed_metric_info[failed_metric.id]["num_failures"] += 1  # type: ignore
+                            failed_metric_info[failed_metric.id]["exception_info"].add(exception_info)  # type: ignore
+                        else:
+                            failed_metric_info[failed_metric.id] = {}
+                            failed_metric_info[failed_metric.id][
+                                "metric_configuration"
+                            ] = failed_metric
+                            failed_metric_info[failed_metric.id]["num_failures"] = 1  # type: ignore
```
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py]: Can you specify what the type error is?

File: great_expectations/validator/validation_graph.py [hunk 2]
```diff
@@ -28,20 +40,190 @@ def id(self):
...
+                    desc="Calculating Metrics",
+                    disable=disable,
+                )
+            progress_bar.update(0)
+            progress_bar.refresh()
+
+            computable_metrics = set()
+
+            for metric in ready_metrics:
+                if metric.id in failed_metric_info and failed_metric_info[metric.id]["num_failures"] >= MAX_METRIC_COMPUTATION_RETRIES:  # type: ignore[operator]  # Incorrect flagging of 'Unsupported operand types for <= ("int" and "MetricConfiguration") and for >= ("Set[ExceptionInfo]" and "int")' in deep "Union" structure.
```
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py:169]: I think this is fixable with some updates to the `MetricConfiguration` operator typing
But I think it's fine to defer this to another PR

File: great_expectations/validator/validation_graph.py [hunk 3]
```diff
@@ -28,20 +40,190 @@ def id(self):
...
+                    exception_traceback = traceback.format_exc()
+                    exception_message = str(err)
+                    exception_info = ExceptionInfo(
+                        exception_traceback=exception_traceback,
+                        exception_message=exception_message,
+                    )
+                    for failed_metric in err.failed_metrics:
+                        if failed_metric.id in failed_metric_info:
+                            failed_metric_info[failed_metric.id]["num_failures"] += 1  # type: ignore[operator]  # Incorrect flagging of 'Unsupported operand types for <= ("int" and "MetricConfiguration") and for >= ("Set[ExceptionInfo]" and "int")' in deep "Union" structure.
+                            failed_metric_info[failed_metric.id]["exception_info"].add(exception_info)  # type: ignore[union-attr]  # Incorrect flagging of 'Item "MetricConfiguration" of "Union[MetricConfiguration, Set[ExceptionInfo], int]" has no attribute "add" and Item "int" of "Union[MetricConfiguration, Set[ExceptionInfo], int]" has no attribute "add"' in deep "Union" structure.
```
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py:196]: I wouldn't say the type system is incorrect here, it's just that we haven't proven to the typing system that this item can't be an `int` etc.

The general term for this is "type narrowing".
https://mypy.readthedocs.io/en/stable/type_narrowing.html?highlight=type%20narrowing#type-narrowing

File: great_expectations/validator/validation_graph.py [hunk 4]
```diff
@@ -28,20 +40,190 @@ def id(self):
...
+                    done = True
+                else:
+                    raise e
+
+            if (len(ready_metrics) + len(needed_metrics) == 0) or (
+                len(ready_metrics) == len(aborted_metrics_info)
+            ):
+                done = True
+
+        progress_bar.close()  # type: ignore[union-attr]  # Incorrect flagging of 'Item "None" of "Optional[Any]" has no attribute "close"' in external package.
```
Comment (Kilo59) [inline great_expectations/validator/validation_graph.py:222]: You could use type-narrowing here to get rid of this warning.

File: great_expectations/validator/validation_graph.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py:169]: @Kilo59 Thanks -- we can address in a follow-on PR.

File: great_expectations/validator/validation_graph.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py:196]: @Kilo59 Thanks -- as agreed, we can address in a follow-on PR.

File: great_expectations/validator/validation_graph.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/validator/validation_graph.py:222]: @Kilo59 Thanks -- as agreed, we can address in a follow-on PR.

**Review by donaldheppner — APPROVED** [review]

**Pull request merged**